mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseToStdlib.cpp

```diff
@@ -239,30 +239,39 @@ struct TransformingForRewritePattern
 //===----------------------------------------------------------------------===//
 // BufferLoad lowering.
 //===----------------------------------------------------------------------===//
-// TODO(kdrewnia): use "OOB reads = 0" from hardware to remove
-// hardcoded zero value
 struct BufferLoadRewritePattern : public OpRewritePattern<BufferLoadOp> {
   using OpRewritePattern<BufferLoadOp>::OpRewritePattern;
   LogicalResult matchAndRewrite(BufferLoadOp op,
                                 PatternRewriter &b) const override {
     Location loc = op.getLoc();
     Value source = op.source();
+    ValueRange coords = op.coords();
     auto sourceType = source.getType().cast<MemRefType>();
     ArrayRef<int64_t> sourceShape = sourceType.getShape();
-    int64_t sourceNumElems = sourceType.getNumElements();
-    SmallVector<int64_t, 5> sourceStrides;
-    int64_t sourceOffset;
-    if (failed(getStridesAndOffset(sourceType, sourceStrides, sourceOffset))) {
-      return op.emitOpError("Somehow we don't have static strides\n");
-    }
 
     Type loadedType = op.result().getType();
-    SmallVector<Value, 5> coords;
-    coords.reserve(op.coords().size());
-    llvm::copy(op.coords(), std::back_inserter(coords));
+    // Emit load instruction
+    // use buffer load since the source memref is on address space 0
+    // If the index here is outside the range of the underlying buffer,
+    // the load will return 0. If indexing produces a final index inside the
+    // buffer but some of the coordinates are out of bounds (ex, loading from an
+    // h or w coordinate that corresponded to implicit padding), correct afther
+    // the fact with a zero
+
+    // This approach allows us to move bounds checks on loads to after each load
+    // which improves performance by reducing memory stalls. It also simplifies
+    // the adress computation logic on loads, enabling potential compiler
+    // optimizations.
```

Comment:
I doubt the impact of address computation logic on loads is as impactful on a gpu. I actually don't know if gpu has branch predictor or how many pipelines it generally has at all.

On occasion of oob, I suspect it will:
 - consume more power because of the additional load
 - pollute l1/l2 cache because of the useless value

But those effect may not be apparent enough in a convolution kernel and again I don't know how much impact to the performance does boudn check has.

Reply:
That would explain the very not-great perf numbers I got here